Issue/master/23 steps1and2 should focus on securing the fleet

jeffmccune opened the pull request:
```
Make steps 1 and 2 only deal with securing the fleet

Without this patch, a new CA is generated in step1.  This new CA may
never get used if the user wishes to stop the process once their entire
fleet has been secured by switching to the new DNS name.

This patch addresses the problem by making some major changes to the way
the process is carried out.

First, steps 1 and 2 are only concerned with securing the Puppet nodes
themselves.  To this end, step1 simply issues a new SSL certificate for
the master with an additional subject alternative name that has not been
issued to anyone else.  Existing agents will continue to work after this
step, however they are still vulnerable until they use the new name.

Step 2 is strictly responsible for configuring all of the agent nodes to
use the new intermediate DNS name.  This is implemented as a Puppet
class named cve20113872::step2.  The step2 script simply adds this class
to the configuration catalog of all nodes by including it at top scope.

The generation of the new CA in step1 (without this patch) is moved into
a discrete step3 (with this patch).  Apache is also re-configured to use
the new CA bundle in step3, causing the master to authenticate agents
who posses certificates issued by both the old and new CA.

This patch _removes_ the functionality of causing agents to generate new
CSR's and submit them to the new CA.  This functionality will be added
back in as step4 as a subsequent patch.  The logic remains in the puppet
class cve20113872::step4 however.

This patch makes the README stale.  The README will be updated in a
subsequent patch.

Closes #18 - The intermediate certificate issued in step1 contains all
previous DNS names now.

Closes #19 - The cve20113872::step2 class is now included at top scope
not in the node scope.
```

Commit:
Make steps 1 and 2 only deal with securing the fleet

Without this patch, a new CA is generated in step1.  This new CA may
never get used if the user wishes to stop the process once their entire
fleet has been secured by switching to the new DNS name.

This patch addresses the problem by making some major changes to the way
the process is carried out.

First, steps 1 and 2 are only concerned with securing the Puppet nodes
themselves.  To this end, step1 simply issues a new SSL certificate for
the master with an additional subject alternative name that has not been
issued to anyone else.  Existing agents will continue to work after this
step, however they are still vulnerable until they use the new name.

Step 2 is strictly responsible for configuring all of the agent nodes to
use the new intermediate DNS name.  This is implemented as a Puppet
class named cve20113872::step2.  The step2 script simply adds this class
to the configuration catalog of all nodes by including it at top scope.

The generation of the new CA in step1 (without this patch) is moved into
a discrete step3 (with this patch).  Apache is also re-configured to use
the new CA bundle in step3, causing the master to authenticate agents
who posses certificates issued by both the old and new CA.

This patch _removes_ the functionality of causing agents to generate new
CSR's and submit them to the new CA.  This functionality will be added
back in as step4 as a subsequent patch.  The logic remains in the puppet
class cve20113872::step4 however.

This patch makes the README stale.  The README will be updated in a
subsequent patch.

Closes #18 - The intermediate certificate issued in step1 contains all
previous DNS names now.

Closes #19 - The cve20113872::step2 class is now included at top scope
not in the node scope.

Commit:
Avoid modifying Apache configuration file

Without this patch, the Apache configuration file is being modified to
point at the bundle of CA certificates and CRL lists once a new CA is
generated.  The user will ultimately migrate solely to the new CA and it
may be confusing for them if we change the configuration half way
through the process then change it back at the end.

This patch avoids the need to modify Apache's configuration by appending
the old CA certificate to the newly generated CA certificate at the
location the Apache configuration points to by default in PE.

Commit:
Add agent_server fact

While agents are being secured in step2, it will be nice if we provide a
way for the end user to check if each node is using the new, secure
intermediate DNS name, or a vulnerable name.

This patch adds the server configuration option to the list of facts the
agent reports to the master.  When the agent is flipped from one DNS
name to another, this may be monitored server side in
yamldir/node/*.yaml

Commit:
Add time stamp to the new CA Common Name

This patch adds a time stamp to the CN field of the newly generated CA
certificate.  This is to avoid name conflicts with any existing or
future CA names.

nigelkersten commented on bin/pe_step1_enable_intermediate_dns_name:
```diff
@@ -0,0 +1,121 @@
+timestamp="$(ruby -e 'puts Time.now.to_i')"
+
+echo -n "Stopping Puppet Master..." >&2
+puppet resource service pe-httpd ensure=stopped hasstatus=true 2>&1 >/dev/null
```
(nitpick) bashisms allowed. &> /dev/null

jeffmccune replied:
Noted, I forgot about &>

jeffmccune replied:
It should be fixed.  I just edited the commit.

Commit:
Improve readability of step3 progress

This commit makes a UX only change to the way progress is presented to
the user.  Each task is printed to STDERR and finished with "done." to
close out the line.

In addition, the puppet commands are hidden from the user.

nigelkersten commented on bin/pe_step2_configure_agents_for_intermediate_dns_name:
```diff
@@ -0,0 +1,75 @@
+# Replace the old master name with the new master name.  This reconfigures the
+# agent on the puppet master to use the intermediate DNS name.
+echo -n "Reconfiguring puppet.conf to connect to the master using intermediate dns name ${new_master_cert} ..." >&2
+ruby -p -l -i.backup.${timestamp}.${idx} -e \
```
Will this cope with people who haven't set the server at all and are relying upon "puppet" ?

I know PE lays down configs with explicit servers, but we can't guarantee that users haven't configured their puppet.conf afterwards.

This is a general comment that applies to everyone of these puppet.conf modifications where we don't check for the existence of the setting explicitly.

jeffmccune replied:
No it won't.

It's actually really difficult to write that logic, especially given the different blocks of `[main]` `[puppetd]` and `[agent]`

Not to mention archive_fileserver and server.

Could we instead ask the user to make sure server is set in puppet.conf before starting the process?

Honestly, we don't _have_ to do this for the agent on the master...  Do we?

nigelkersten replied:
As said in person, I don't think the logic to detect if the parameter is set is difficult.

```
puppet agent --genconfig |grep ' server ='
    # server = puppet
```

This will parse main,agent,puppetd,whatever blocks and tell you the result. If it has a # you know that it's not set in the config file and we can just insert it, if it doesn't, then we just gsub it.

nigelkersten replied:
(closing off this loop of comments, will file an issue around the larger problem)
